Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -0,0 +1,479 @@
+                std::vector<amrex::ParticleReal> curr(numParticleOnTile, 0);
+                for (auto i=0; i<numParticleOnTile; i++) {
+                     curr[i] = aos[i].m_rdata.pos[currDim];
+                }
+                currSpecies["position"][axisNames[currDim]].storeChunk(curr, {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+                m_Series->flush();
+           }
+        }
+         //  save properties
+         SaveRealProperty(pti,
+             currSpecies,
+             offset,
+             write_real_comp, real_comp_names);
+
+         offset += numParticleOnTile;
+      }
+    }
+}
+
+void
+WarpXOpenPMDPlot::SetupRealProperties(openPMD::ParticleSpecies& currSpecies,
+                      const amrex::Vector<int>& write_real_comp,
+                      const amrex::Vector<std::string>& real_comp_names,
+                      unsigned long long np) const
+{
+  auto particlesLineup = openPMD::Dataset(openPMD::determineDatatype<amrex::ParticleReal>(), {np});
+
+  //
+  // the beam/input3d showed write_real_comp.size() = 16 while only 10 real comp names
+  // so using the min to be safe.
+  //
+  auto counter = std::min(write_real_comp.size(), real_comp_names.size());
+  for (int i = 0; i < counter; ++i)
+    if (write_real_comp[i]) {
+      auto& particleVar = currSpecies[real_comp_names[i]];
+      auto& particleVarComp = particleVar[openPMD::RecordComponent::SCALAR];
+      particleVarComp.resetDataset(particlesLineup);
+    }
+}
+
+void
+WarpXOpenPMDPlot::SaveRealProperty(WarpXParIter& pti,
+                       openPMD::ParticleSpecies& currSpecies,
+                       unsigned long long offset,
+                       const amrex::Vector<int>& write_real_comp,
+                       const amrex::Vector<std::string>& real_comp_names) const
+
+{
+  int numOutputReal = 0;
+  int totalRealAttrs = m_NumAoSRealAttributes + m_NumSoARealAttributes;
+
+  for (int i = 0; i < totalRealAttrs; ++i)
+    if (write_real_comp[i])
+      ++numOutputReal;
+
+  auto numParticleOnTile = pti.numParticles();
+  const auto& aos = pti.GetArrayOfStructs();  // size =  numParticlesOnTile
+  const auto& soa = pti.GetStructOfArrays();
+
+  // properties are saved seperately
+  {
+    for (auto idx=0; idx<m_NumAoSRealAttributes; idx++) {
+      if (write_real_comp[idx]) {
+          auto& currVar = currSpecies[real_comp_names[idx]][openPMD::RecordComponent::SCALAR];
+          typename amrex::ParticleReal *d =
+                 static_cast<typename amrex::ParticleReal*> (malloc(sizeof(typename amrex::ParticleReal) *  numParticleOnTile));
+
+          for (auto kk=0; kk<numParticleOnTile; kk++)
+               d[kk] = aos[kk].m_rdata.arr[AMREX_SPACEDIM+idx];
+
+          std::shared_ptr <typename amrex::ParticleReal> data(d, free);
+          currVar.storeChunk(data,
+               {offset}, {static_cast<unsigned long long>(numParticleOnTile)});
+          m_Series->flush();
```

Comment:
Since you are malloc-ing new data, one could potentially also flush multiple variables together at a later point, e.g. after the for loop (or even just at the end of the function).
This will increase the memory footprint on the host side (usually fine) and potentially increases I/O performance.

Reply:
For this one, variable is written immediately after storeChunk. 
This will likely be revisited at the later point.